### ISSUE TITLE
Schedulers.

### ISSUE THREAD

**Issue by GeneralAwareness** [kohya-ss/sd-scripts#217]
I really, really like this scheduler because it gives us a minimum LR that we will not go below.

![image](https://user-images.githubusercontent.com/124339252/220466446-536fb589-2408-4e52-83eb-daf2e89b8f2d.png)

Any chance we can get this implemented?  There are a lot of schedulers that seem to be missing.

**Comment by slashedstar** [comment]
This one has this, plus a lot of other stuff (steps for the first cycle, cycle length multiplier, etc)
https://github.com/katsura-jp/pytorch-cosine-annealing-with-warmup

**Comment by GeneralAwareness** [comment]
> This one has this, plus a lot of other stuff (steps for the first cycle, cycle length multiplier, etc) https://github.com/katsura-jp/pytorch-cosine-annealing-with-warmup

How do I use that on colab?

**Comment by GeneralAwareness** [comment]
> The script currently using schedulers from `Diffusers`, and there is no option for minimum LR. I will add a minimum LR option for cosine with restarts.

Can't you add the cosineAnnealing because I really dislike the restarts?

**Comment by slashedstar** [comment]
> Could you add the cosineAnnealing because I really dislike the restarts as it restarting (jumps back to full LR) really messed with my dataset when I tried it?

You can set restarts to 1 to make it not restart at all, or use just the cosine scheduler (which should have no restarts)

**Comment by GeneralAwareness** [comment]
Yes, but he said he was only adding it to the cosine with restarts.  btw, I was reading about this and we are missing some things from it.

**Comment by AI-Casanova** [comment]
> 1 would be to restart at each epoch, 2 after 2 epochs, 3 after 3 epochs, and so on and so on.

That would be "Restart every n epochs" not "# of restarts"

The one you're proposing is the torch method, the one in code is the diffusers method.

Perhaps that could be made more clear, but the terminology is correct for the optimizer used.

**Comment by AI-Casanova** [comment]
@GeneralAwareness 

1 restart should be 2 full cycles. 

# restarts = # epochs would be torch 1

In other words, diffusers # = epochs / torch #

![Uploading Screenshot_20230223-143527-732.png…]()

**Comment by GeneralAwareness** [comment]
@AI-Casanova Huh?    4 epochs for a total training in the torch world would be 4 restarts.  In the diffusers version that would need to be 4 or 0.25 (in this case)?  If 0.25 then I was right in the maths, and if not ugh.

**Comment by GeneralAwareness** [comment]
Was this fixed? https://github.com/huggingface/diffusers/issues/1232

**Comment by AI-Casanova** [comment]
Oh, maybe they are offset by 1 🤔

**Comment by GeneralAwareness** [comment]
@AI-Casanova I am not sure tbh, but I really wonder if the issue I posted from HuggingFace's implementation of this has been fixed as it was actually not restarting before.

**Comment by GeneralAwareness** [comment]
I received a reply about this as such @kohya-ss   Is this implemented in yours?

![image](https://user-images.githubusercontent.com/124339252/221330272-96b6b91d-f709-41e5-8684-eeff83585a76.png)
